### PR TITLE
Hotfix/key thrown

### LINKED ISSUE #90: undefined is already in use on file upload
image
This error appears sometimes randomly when uploading a file (bigger or smaller than 5MB) to the application and its returned from the file-service.

### PR DESCRIPTION
**Pull request by Shahar-Y**
Changed the way duplicate key errors are treated, using regex. Now the errors are more precise.
Fixed the option of the error showing undefined (#90)
Manually tested with docker, it works.
Closes #90, #107

### CODE REVIEW

File: src/file/file.model.ts [hunk 1]
```diff
@@ -76,14 +76,31 @@ fileSchema.virtual('fullExtension')
...
-    next(new KeyAlreadyExistsError(this.key));
+    const retMessage : string = getMongoErrorIndices(error);
+    next(new KeyAlreadyExistsError(retMessage));
   } else {
     next();
   }
 };
 
+function getMongoErrorIndices(error: MongoError) : string {
+  const indicesRegex : RegExp = new RegExp(/index\:\ (?:.*\.)?\$?(?:([_a-z0-9]*)(?:_\d*)|([_a-z0-9]*))\s*dup key/i);
```
Comment (omerzamir) [inline src/file/file.model.ts]: WTF is this?

File: src/file/file.model.ts [hunk 2]
```diff
@@ -76,14 +76,31 @@ fileSchema.virtual('fullExtension')
...
+  const indicesMatch : RegExpMatchArray =  error.message.match(indicesRegex);
+  let indexName : string = indicesMatch[1] || indicesMatch[2];
+
+  const re : RegExp = new RegExp('_1_', 'g');
```
Comment (omerzamir) [inline src/file/file.model.ts]: WTF is this?

File: src/file/file.model.ts [hunk 3]
```diff
@@ -76,14 +76,31 @@ fileSchema.virtual('fullExtension')
...
+  indexName = indexName.replace(re, ', ');
+
+  const valuesRE : RegExp = new RegExp(/{(.*?)}/);
```
Comment (omerzamir) [inline src/file/file.model.ts]: WTF is this?

File: src/file/file.model.ts [hunk 4]
```diff
@@ -76,14 +76,31 @@ fileSchema.virtual('fullExtension')
...
+  const valuesMatch : RegExpMatchArray = error.message.match(valuesRE);
+  let values : string = valuesMatch[0];
+  values = values.replace(new RegExp(' : ', 'g'), ' ');
```
Comment (omerzamir) [inline src/file/file.model.ts]: WTF is this?

**Comment by Shahar-Y** [comment]
The regex is taken from this mongoose issue: https://github.com/Automattic/mongoose/issues/2129 
Checked it both with a single index an multiple indices

File: src/file/file.model.ts (hunk 1 above)
Comment (Shahar-Y) [reply src/file/file.model.ts]: Added documentation

File: src/file/file.model.ts [hunk 5]
```diff
@@ -76,14 +76,39 @@ fileSchema.virtual('fullExtension')
   });
 
 // handleE11000 is called when there is a duplicateKey Error
-const handleE11000 = function (error: MongoError, _: any, next: NextFunction) {
+const handleE11000 = function (error: MongoError, _: any, next: NextFunction) : void {
   if (error.name === 'MongoError' && error.code === 11000) {
-    next(new KeyAlreadyExistsError(this.key));
+    const retMessage : string = getMongoErrorIndices(error);
+    next(new KeyAlreadyExistsError(retMessage));
```
Comment (AllKinds) [inline src/file/file.model.ts]: Maybe we should create a different error for cases where the compound indexes already exist. We can accidentally confuse it with the key field.

File: src/file/file.service.spec.ts [hunk 6]
```diff
@@ -357,14 +357,37 @@ describe('File Logic', () => {
...
-    it('should throw an error when KEY already exist', async () => {
+    it('should throw an error when KEY already exists', async () => {
       await FileService.create(bucket, 'tmp1', USER.id, 'text', null, KEY);
       await FileService.create(bucket, 'tmp2', USER.id, 'text', null, KEY)
       .should.eventually.be.rejectedWith(KeyAlreadyExistsError);
     });
 
   });
 
+  describe('#updateById', () => {
```
Comment (AllKinds) [inline src/file/file.service.spec.ts]: How does updateById tests related to this issue/hotfix?

File: src/file/file.service.spec.ts [hunk 7]
```diff
@@ -357,14 +357,37 @@ describe('File Logic', () => {
       expect(file2.parent).to.equal(file1.parent);
     });
 
-    it('should throw an error when KEY already exist', async () => {
+    it('should throw an error when KEY already exists', async () => {
```
Comment (AllKinds) [inline src/file/file.service.spec.ts]: Why did you delete the `s` in `exists`?

File: src/file/file.model.ts (hunk 5 above)
Comment (Shahar-Y) [reply src/file/file.model.ts]: Changed to 'UniqueIndexExistsError'

File: src/file/file.service.spec.ts (hunk 6 above)
Comment (Shahar-Y) [reply src/file/file.service.spec.ts]: I created tests that were missing and are related to this issue

File: src/file/file.model.ts (hunk 2 above)
Comment (Shahar-Y) [reply src/file/file.model.ts]: Added documentation

File: src/file/file.model.ts (hunk 3 above)
Comment (Shahar-Y) [reply src/file/file.model.ts]: Added documentation

File: src/file/file.model.ts (hunk 4 above)
Comment (Shahar-Y) [reply src/file/file.model.ts]: Added documentation

Comment (Shahar-Y) [reply src/file/file.model.ts]: Added documentation
